libmamba/include/mamba/util/iterator.hpp

```diff
@@ -0,0 +1,400 @@
+                                               xbidirectional_iterator_base_from_traits<I, T>>;
+    };
+
+    template <class Iterator, class R = void>
+    using enable_random_access_iterator = std::enable_if_t<detail::is_random_access_iterator<Iterator>::value, R>;
+
+    /*******************
+     * filter_iterator *
+     *******************/
+
+    template <class Predicate, class Iterator>
+    class filter_iterator;
+
+    template <class Predicate, class Iterator>
+    struct select_filter_iterator_base
+    {
+        using type = typename select_iterator_base<Iterator>::from_traits<
+            filter_iterator<Predicate, Iterator>,
+            std::iterator_traits<Iterator>
+        >;
+    };
+
+    template <class Predicate, class Iterator>
+    using select_filter_iterator_base_t = typename select_filter_iterator_base<Predicate, Iterator>::type;
+
+    template <class Predicate, class Iterator>
+    class filter_iterator : public select_filter_iterator_base_t<Predicate, Iterator>
+    {
+    public:
+
+        using self_type = filter_iterator<Predicate, Iterator>;
+        using base_type = select_filter_iterator_base_t<Predicate, Iterator>;
+
+        using reference = typename base_type::reference;
+        using pointer = typename base_type::pointer;
+        using difference_type = typename base_type::difference_type;
+        using iterator_category = typename base_type::iterator_category;
+
+        filter_iterator() = default;
+
+        filter_iterator(Iterator iter, Iterator begin, Iterator end)
+            : m_pred()
+            , m_iter(iter)
+            , m_begin_limit(--begin)
+            , m_end(end)
+        {
+            satisfy_predicate();
+        }
+
+        template <class Pred>
+        filter_iterator(Pred&& pred, Iterator iter, Iterator begin, Iterator end)
+            : m_pred(std::forward<Pred>(pred))
+            , m_iter(iter)
+            , m_begin_limit(--begin)
+            , m_end(end)
+        {
+            satisfy_predicate();
+        }
+
+        self_type& operator++()
+        {
+            ++m_iter;
+            satisfy_predicate();
+            return *this;
+        }
+
+        self_type& operator--()
+        {
+            --m_iter;
+            while (m_iter != m_begin_limit && !m_pred(*m_iter))
+            {
+                --m_iter;
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, self_type>
+        operator+=(difference_type n)
+        {
+            if (n >= 0)
+            {
+                advance(n);
+            }
+            else
+            {
+                step_back(n);
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, self_type>
+        operator-=(difference_type n)
+        {
+            if (n >= 0)
```

Comment:
As mentionned below, this check of direction should be done in `advance` which should handle both directions, also matchnig how `std::advance` does.